File: plugins/plugin-nostr/typescript/src/providers/senderContext.ts
/**
 * Sender context provider for Nostr plugin.
 */

import type {
  IAgentRuntime,
  Memory,
  Provider,
  ProviderResult,
  State,
} from "@elizaos/core";
import type { NostrService } from "../service.js";
import {
  getPubkeyDisplayName,
  NOSTR_SERVICE_NAME,
  pubkeyToNpub,
} from "../types.js";

export const senderContextProvider: Provider = {
  name: "nostrSenderContext",
  description:
    "Provides information about the Nostr user in the current conversation",

  get: async (
    runtime: IAgentRuntime,
    message: Memory,
    state: State,
  ): Promise<ProviderResult> => {
    // Only provide context for Nostr messages
    if (message.content.source !== "nostr") {
      return {
        data: {},
        values: {},
        text: "",
      };
    }

    const nostrService = await runtime.getService<NostrService>(NOSTR_SERVICE_NAME);

    if (!nostrService || !nostrService.isConnected()) {
      return {
        data: { connected: false },
        values: { connected: false },
        text: "",
      };
    }

    const agentName = state?.agentName || "The agent";

    // Get sender pubkey from state if available
    const senderPubkey = state?.data?.senderPubkey as string | undefined;

    if (!senderPubkey) {
      return {
        data: { connected: true },
        values: { connected: true },
        text: "",
      };
    }

    let senderNpub = "";
    try {
      senderNpub = pubkeyToNpub(senderPubkey);
    } catch {
      // Use hex if npub conversion fails
    }

    const displayName = getPubkeyDisplayName(senderPubkey);

    const responseText =
      `${agentName} is talking to ${displayName} on Nostr. ` +
      `Their pubkey is ${senderNpub || senderPubkey}. ` +
      `This is an encrypted direct message conversation using NIP-04.`;

    return {
      data: {
        senderPubkey,
        senderNpub,
        displayName,
        isEncrypted: true,
      },
      values: {
        senderPubkey,
        senderNpub,
        displayName,
      },
      text: responseText,
    };
  },
};

File: plugins/plugin-nostr/rust/src/providers/sender_context.rs
//! Sender context provider for the Nostr plugin.

use crate::service::NostrService;
use crate::types::{get_pubkey_display_name, pubkey_to_npub};
use serde::{Deserialize, Serialize};

/// Sender context data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NostrSenderContext {
    pub sender_pubkey: String,
    pub sender_npub: String,
    pub display_name: String,
    pub is_encrypted: bool,
}

/// Sender context response
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SenderContextResponse {
    pub data: Option<NostrSenderContext>,
    pub values: serde_json::Value,
    pub text: String,
}

/// Get the Nostr sender context
pub async fn get_sender_context(
    service: &NostrService,
    sender_pubkey: Option<&str>,
    agent_name: Option<&str>,
) -> SenderContextResponse {
    let agent = agent_name.unwrap_or("The agent");

    if !service.is_connected().await {
        return SenderContextResponse {
            data: None,
            values: serde_json::json!({ "connected": false }),
            text: String::new(),
        };
    }

    let pubkey = match sender_pubkey {
        Some(pk) => pk,
        None => {
            return SenderContextResponse {
                data: None,
                values: serde_json::json!({ "connected": true }),
                text: String::new(),
            }
        }
    };

    let sender_npub = pubkey_to_npub(pubkey).unwrap_or_default();
    let display_name = get_pubkey_display_name(pubkey);

    let text = format!(
        "{} is talking to {} on Nostr. Their pubkey is {}. \
         This is an encrypted direct message conversation using NIP-04.",
        agent,
        display_name,
        if sender_npub.is_empty() {
            pubkey.to_string()
        } else {
            sender_npub.clone()
        }
    );

    SenderContextResponse {
        data: Some(NostrSenderContext {
            sender_pubkey: pubkey.to_string(),
            sender_npub: sender_npub.clone(),
            display_name: display_name.clone(),
            is_encrypted: true,
        }),
        values: serde_json::json!({
            "sender_pubkey": pubkey,
            "sender_npub": sender_npub,
            "display_name": display_name,
        }),
        text,
    }
}

/// Provider metadata
pub const SENDER_CONTEXT_PROVIDER_NAME: &str = "nostrSenderContext";
pub const SENDER_CONTEXT_PROVIDER_DESCRIPTION: &str =
    "Provides information about the Nostr user in the current conversation";
